### DISCUSSION TITLE
Parser

### DISCUSSION THREAD

**Discussion by alensiljak** [jcornaz/beancount_parser_2#26] [General]
Hi! Apologies that my messages will be brief, I'm typing on a phone so please bear with me. 

I just had a brief look (also on the phone 🙈) at the implementation and notice that you're using Nom for parsing. Do you mind sharing some findings as to the ease of implementation, the performance, etc.? 

In the ledger-rs initiative, I've looked at probably all the parsers available on crates.io and in the end went ahead with creating a custom parser, borrowing the logic directly from Ledger. 
Some performance tests done as a part of the project showed the original Ledger running laps around any of the generic Rust parsers. 

So I'm interested in your experiences.

**Comment by jcornaz** [comment]
> Do you mind sharing some findings as to the ease of implementation, the performance, etc.?

No, I don't mind. Long story short I like nom, and I'd recommend it to anyone who's serious about building a parser.
My only comparison is [pest](pest.rs), or building a parser myself from scratch. so here's how I personally compare pest and nom:

* `nom` is a little bit harder to learn at first. But once you get the concept and know the [main combinators](https://github.com/rust-bakery/nom/blob/main/doc/choosing_a_combinator.md) then I find it very easy to use, maybe even easier than `pest`.
* When it comes to performance, nom is significantly faster than `pest`. At least that's what I found in my (flawed) [benchmark](https://github.com/jcornaz/beancount-parser/blob/main/benches/benchmark.rs) of `beancount-parser`. But no matter how flawed my benchmark is https://pest.rs also admits to being slower than nom. And I am personally not skilled enough to build a parser from scratch faster than `nom`.
* `pest` however delivers a nicer end-user experience out of the box, especially when it comes to error messages. I have yet a lot of work to make my error as useful as they are with `pest` out of the box.

When it comes to building a parser from scratch, I anyway tend to have an approach that's very similar to nom (each function returns the remaining input to parse alongside the success or error result).
As fun as building a parser from scratch is, if I have anything complex to parse (like ledger or beancount) then I'd go for nom.

A small gotcha with nom, I'd recommend using [nom_locate](https://github.com/fflorent/nom_locate) from the get-go. Or at least don't wait too long before introducing it. As nom doesn't track position in the source input out of the box.

**Reply by jcornaz** [reply]
Maybe something else to note: `pest` publishes breaking changes in their patch versions, which is a dealbreaker for me :-/

I actually just learned that 2 days ago, when my former `beancount-parser` project stopped building since the publication of pest `2.6.1`. And it was apparently not a mistake, they actually mention it in a section name "Warning: Semantic Versioning" of their release note: https://github.com/pest-parser/pest/releases/tag/v2.6.1

I haven't investigated exactly what they changed, I decided that it was the last nail in `pest`'s coffin and I removed my alternative parser implementation which was using pest. Personally, I'll be sticking with `nom` moving forward.

**Reply by alensiljak** [reply]
Yes, nom seems interesting. It is exactly what one writes when writing a custom parser. When I was starting, having no prior experience and a choice of options to try out, I focused on pest since I preferred to have the grammar separated.
Later I went ahead with borrowing Ledger rules directly. Now, having some understanding of the logic, I think it would be interesting to convert those custom rules to nom. I anyway separated the scanner from the parser, the parser being the code that creates the data structure of the domain model, so I assume the conversion would be straightforward.
The main outcome being the knowledge and experience for the project.
